### PR TITLE
Update: Installation and setup

### LINKED ISSUE #57: Installation and setup
**Is your enhancement related to a problem? Please describe.**
Better onboarding and easier setup - leading to better initial experience.  Better control on behaviour:
- Ability to define which post types will Autoshare be available on
- Configurable default ON/OFF for posts
- Ability to configure default featured image inclusion
- Ability to tweet a post multiple times

**Describe the solution you'd like**
- [ ] Prompt what the user needs to do after Activating the plugin (need to decide if Admin banner is best, necessary UX here)
- [ ] Provide step-by-step instructions
- [ ] Only show settings which are needed for initial setup
- [ ] Branding and links to more in-depth help
- [ ] Group configuration options by areas
- [ ] Only show post types when "Selected content types only" is selected
- [ ] Toggle connection settings inline

**Designs**
Activating the plugin:
Screen Shot 2019-12-13 at 7 26 22 PM
Initial setup:
Screen Shot 2019-12-13 at 7 28 25 PM
Configuration:
Screen Shot 2019-12-13 at 7 29 37 PM
Configuration - open states:
Screen Shot 2019-12-13 at 7 33 20 PM

**Describe alternatives you've considered**
n/a

**Additional context**
Props to @linawiezkowiak and @oszkarnagy for the UX/Design work.

### PR DESCRIPTION
**Pull request by dinhtungdu**
### Description of the Change
Update and improve the setting page to make the plugin easier to set up and use. See #57.

### Alternate Designs
n/a

### Benefits
Setting up our plugin will be easier and less frustrated to use than current.

### Possible Drawbacks
n/a

### Checklist:

- [x] I have read the [**CONTRIBUTING**](/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my change.
- [x] All new and existing tests passed.

### Applicable Issues
Close #57

### Changelog Entry

* Updated: New and improved setting page.

### CODE REVIEW

File: includes/admin/settings.php [hunk 1]
```diff
@@ -133,17 +294,54 @@ function options_page() {
...
+	if ( Utils\is_twitter_configured() ) {
+		$links['settings'] = sprintf(
+			/* translators: %s is the plugin setting page URL */
+			__( '<a href="%s">Settings</a>', 'autoshare-for-twitter' ),
+			esc_url( admin_url( 'options-general.php?page=autoshare-for-twitter' ) )
+		);
+	} else {
+		$links['initial-setup'] = sprintf(
+			/* translators: %s is the plugin setting page URL */
+			__( '<a href="%s">Setup your Twitter account</a>', 'autoshare-for-twitter' ),
```
Comment (jeffpaul) [inline includes/admin/settings.php]: Split `Setup` into `Set up`

File: includes/admin/settings.php [hunk 2]
```diff
@@ -133,17 +294,54 @@ function options_page() {
...
+			<div class="brand">
+				<a href="https://10up.com" class="logo" title="<?php echo esc_html__( '10up', 'autoshare-for-twitter' ); ?>">
+					<img src="<?php echo esc_url( trailingslashit( AUTOSHARE_FOR_TWITTER_URL ) . 'assets/images/10up.svg' ); ?>" alt="<?php echo esc_html__( '10up logo', 'autoshare-for-twitter' ); ?>" />
+				</a>
+				<p>
+					<strong>
+						<?php echo esc_html__( 'Autoshare for Twitter', 'autoshare-for-twitter' ) . ' ' . esc_html__( 'by', 'autoshare-for-twitter' ); ?> <a href="https://10up.com" class="logo" title="<?php echo esc_html__( '10up', 'autoshare-for-twitter' ); ?>"><?php echo esc_html__( '10up', 'autoshare-for-twitter' ); ?></a>
+					</strong>
+				</p>
+			</div>
```
Comment (jeffpaul) [inline includes/admin/settings.php]: Let's also include a link for `Support` underneath this line that links out to https://github.com/10up/autoshare-for-twitter/issues

**Comment by jeffpaul** [comment]
Let's update the initial set up flow to be more like https://user-images.githubusercontent.com/2818133/70841275-c5669100-1dde-11ea-87de-36b212d411e6.png where we require folks to establish their connection to Twitter before getting into plugin-specific settings.  Once their connection to Twitter is setup, we can get into plugin settings like https://user-images.githubusercontent.com/2818133/70841299-ee872180-1dde-11ea-8229-763fc80f3b21.png and https://user-images.githubusercontent.com/2818133/70841348-73723b00-1ddf-11ea-90e3-fd0de332893f.png.  The following copy can be used for this initial settings page...

# Autoshare for Twitter settings

## Twitter connection settings

### 1. [Apply for a Twitter developer account](https://developer.twitter.com/en/apply/user.html)

### 2. Create a Twitter app for your website

- Click the `Create an app` button on the [Twitter develop apps page](https://developer.twitter.com/en/apps).
- Fill out the `App name` and `Application description` fields.
- Set the `Website URL` and `Callback URLs` fields to **https://yourdomain.yourdomainextension**.
- Fill out the `Tell us how this app will be used` field, no other fields or URLs are required or necessary.

### 3. Configure access to your Twitter app's API keys

- Click on the `Keys and tokens` tab within your newly created Twitter developer app.
- Copy the `API key` and `API secret key` values from your Twitter app `Consumer API keys` section and paste them below.
API key [`paste your API key here`]
API secret key [`paste your API secret key here`]

### 4. Configure access to your Twitter app's access tokens

- Click on the `Generate` button from your Twitter app `Access token & access token secret` section.
- Copy the `Access token` and `Access token secret` values and paste them below.
Access token [`paste your Access token here`]
Access token secret [`paste your Access token secret here`]

### 5. Confirm Twitter handle

- Fill out your Twitter handle that will be used to tweet your posts, pages, etc.
Twitter handle [`enter your Twitter handle here`]

### 6. Connect your Twitter developer app with this site

- Click the `Connect to Twitter` button below 🎉 
[`Connect to Twitter`]

**Comment by jeffpaul** [comment]
After someone clicks the `Connect to Twitter` button from the initial plugin settings, the plugin settings page should save those details / connect to Twitter, and redirect to the post-Twitter connection settings/Autoshare settings version of the plugin settings page like https://user-images.githubusercontent.com/2818133/70841299-ee872180-1dde-11ea-8229-763fc80f3b21.png.  The following copy can be used for this initial settings page...

# Autoshare for Twitter settings

## Tweet settings

Enable Autoshare for:
(*) All content types <-- this should be our default
( ) Selected content types only
[ ] Posts
[ ] Pages
[ ] <CPT Name>  <-- we'll want to include checkboxes for any CPTs registered on the site

[x] Enable Autoshare by default when publishing content   <-- enabled by default
[x] Add featured image to the tweet by default   <-- enabled by default

## Twitter connection settings

*Open connection settings \\/*  <-- clicking this expands to show the full Twitter connection settings as described in the comment above and like https://user-images.githubusercontent.com/2818133/70841348-73723b00-1ddf-11ea-90e3-fd0de332893f.png

[`Update settings`]

**Comment by dinhtungdu** [comment]
> (*) All content types <-- this should be our default

I think we should keep the current behavior of our plugin that supports only `post` and `page` by default.

> [ ] <-- we'll want to include checkboxes for any CPTs registered on the site

Yes, I include all custom post types to the post type checkbox. Also, if any post type is enabled using code, the checkbox will be checked and disabled.

**Comment by jeffpaul** [comment]
@dinhtungdu 

> (*) All content types <-- this should be our default

> I think we should keep the current behavior of our plugin that supports only post and page by default.

Seems good, agreed.

File: includes/admin/settings.php [hunk 3]
```diff
@@ -133,17 +363,65 @@ function options_page() {
...
+				<p>
+					<strong>
+						<?php echo esc_html__( 'Autoshare for Twitter', 'autoshare-for-twitter' ) . ' ' . esc_html__( 'by', 'autoshare-for-twitter' ); ?> <a href="https://10up.com" class="logo" title="<?php echo esc_html__( '10up', 'autoshare-for-twitter' ); ?>"><?php echo esc_html__( '10up', 'autoshare-for-twitter' ); ?></a>
+					</strong>
+				</p>
+				<nav>
+					<a href="https://github.com/10up/autoshare-for-twitter#installation" target="_blank" title="<?php echo esc_attr__( 'Installation & Set up', 'autoshare-for-twitter' ); ?>">
+						<?php echo esc_html__( 'Installation & Set up', 'autoshare-for-twitter' ); ?><span class="dashicons dashicons-external"></span>
+					</a>
+					<a href="https://github.com/10up/autoshare-for-twitter#faq" target="_blank" title="<?php echo esc_attr__( 'FAQ', 'autoshare-for-twitter' ); ?>">
```
Comment (jeffpaul) [inline includes/admin/settings.php]: Current link is plural "faqs"

File: includes/admin/settings.php [hunk 4]
```diff
@@ -133,17 +363,65 @@ function options_page() {
...
+				<a href="https://10up.com" class="logo" title="<?php echo esc_html__( '10up', 'autoshare-for-twitter' ); ?>">
+					<img src="<?php echo esc_url( trailingslashit( AUTOSHARE_FOR_TWITTER_URL ) . 'assets/images/10up.svg' ); ?>" alt="<?php echo esc_html__( '10up logo', 'autoshare-for-twitter' ); ?>" />
+				</a>
+				<p>
+					<strong>
+						<?php echo esc_html__( 'Autoshare for Twitter', 'autoshare-for-twitter' ) . ' ' . esc_html__( 'by', 'autoshare-for-twitter' ); ?> <a href="https://10up.com" class="logo" title="<?php echo esc_html__( '10up', 'autoshare-for-twitter' ); ?>"><?php echo esc_html__( '10up', 'autoshare-for-twitter' ); ?></a>
+					</strong>
+				</p>
+				<nav>
+					<a href="https://github.com/10up/autoshare-for-twitter#installation" target="_blank" title="<?php echo esc_attr__( 'Installation & Set up', 'autoshare-for-twitter' ); ?>">
```
Comment (jeffpaul) [inline includes/admin/settings.php]: Probably can remove this link as the plugin is already installed and we're adding set up instructions via this PR

**Comment by jeffpaul** [comment]
@dinhtungdu some UI updates for you...
- [ ] please update the settings page title from `Autoshare for Twitter` to `Autoshare for Twitter Settings`
- [ ] please update from double quotes "" around items to `preformatted text/code` format
- [ ] update yourdomain.yourdomainextension to **https://yourdomain.yourdomainextension**
- [ ] move the `API key` and `API secret key` labels+fields up into section 3
- [ ] add `paste your API key here` as helper text in the `API key` field
- [ ] add `paste your API secret key here` as helper text in the `API secret key` field
- [ ] remove duplicate section 3
- [ ] move the `Access token` and `Access token secret` labels+fields up into section 4
- [ ] add `paste your Access token here` as helper text in the `Access token` field
- [ ] add `paste your Access token secret here` as helper text in the `Access token secret` field
- [ ] move the `Twitter handle` and `https://twitter.com/` labels+field up into section 5
- [ ] add `enter your Twitter handle here` as helper text in the `Twitter handle` field
- [ ] on initial settings connection, have the button label be `Connect to Twitter` and not `Save Changes`
- [ ] after connecting to Twitter, ensure the `Selected content types only` radio button is selected by default (right now nothing is selected by default)
- [ ] change link label to `Open connection settings` from `Open credentials settings`
- [ ] when the connection settings are expanded, we'll want to continue displaying the 1-6 steps
- [ ] after connecting to Twitter, have the button label be `Update Settings` instead of `Save Changes`
- [ ] I'm starting to think we don't need the `Enable Autoshare by default when publishing content` setting because I don't think we have a toggle on/off in the content publishing workflow so by nature if the plugin is active then autoshare is enabled; if you agree, then let's remove that setting (as I don't know that it's tied to any action in the plugin currently)
- [ ] change `Upload featured image` to `Image setting`
- [ ] change ` Add featured image to the tweet by default` to `Always add the featured image to tweets`

Let me know if any of these are confusing or need further explanation, thanks!

**Comment by jeffpaul** [comment]
After chatting with @dinhtungdu, we decided to strike the moving fields and button label changes as too hacky and not a simple solution that otherwise our plugin does a good job of staying simple.

**Comment by jeffpaul** [comment]
Also noting from our discussion that it makes sense to keep the `Enable by default` setting, so we struck that from the UI change list as well.

File: includes/admin/settings.php [hunk 5]
```diff
@@ -113,12 +202,156 @@ function register_settings() {
...
+			</a>
+			<a href="JavaScript:void(0);" class="close">
+				<?php echo esc_html__( 'Close connection settings', 'autoshare-for-twitter' ); ?><span class="dashicons dashicons-arrow-up-alt2"></span>
+			</a>
+		</p>
+		<section class="credentials-instructions">
+			<h4><a href="https://developer.twitter.com/en/apply/user.html" target="_blank"><?php echo esc_html__( '1. Apply for a Twitter developer account', 'autoshare-for-twitter' ); ?></a></h4>
+			<h4><?php esc_html_e( '2. Create a Twitter app for your website', 'autoshare-for-twitter' ); ?></h4>
+			<ul>
+				<li><?php echo wp_kses_data( __( 'Click the <code>Create an app</code> button on the', 'autoshare-for-twitter' ) ); ?> <a href="https://developer.twitter.com/en/apps" title="<?php esc_html_e( 'Twitter develop apps page.', 'autoshare-for-twitter' ); ?>"><?php esc_html_e( 'Twitter develop apps page.', 'autoshare-for-twitter' ); ?></a></li>
```
Comment (jeffpaul) [inline includes/admin/settings.php]: Please update `develop` to `developer`.

File: includes/admin/settings.php [hunk 6]
```diff
@@ -79,7 +157,11 @@ function register_settings() {
...
 		'autoshare-cred_section',
-		[ 'name' => 'access_token' ]
+		[
+			'name'        => 'access_token',
+			'class'       => 'large-text',
+			'placeholder' => __( 'paste your Access token here', 'autoshare-for-twitter' ),
+		]
 	);
 
 	// Access Secret.
```
Comment (jeffpaul) [inline includes/admin/settings.php]: Please update the field label from `Access secret` to `Access token secret`.

File: includes/admin/settings.php [hunk 7]
```diff
@@ -113,12 +202,156 @@ function register_settings() {
...
+				<li><?php echo wp_kses_data( __( 'Click on the <code>Generate</code> button from your Twitter app <code>Access token & access token secret</code> section.', 'autoshare-for-twitter' ) ); ?></li>
+				<li><?php echo wp_kses_data( __( 'Copy the <code>Access token</code> and <code>Access token secret</code> values and paste them below.', 'autoshare-for-twitter' ) ); ?></li>
+			</ul>
+			<h4><?php esc_html_e( '5. Confirm Twitter handle', 'autoshare-for-twitter' ); ?></h4>
+			<ul>
+				<li><?php esc_html_e( 'Fill out your Twitter handle that will be used to tweet your posts, pages, etc.', 'autoshare-for-twitter' ); ?></li>
+			</ul>
+			<h4><?php esc_html_e( '6. Connect your Twitter developer app with this site', 'autoshare-for-twitter' ); ?></h4>
+			<ul>
+				<li><?php echo wp_kses_data( __( 'Click the <code>Connect to Twitter</code> button below.', 'autoshare-for-twitter' ) ); ?></li>
```
Comment (jeffpaul) [inline includes/admin/settings.php]: The button label is `Save Changes` so if we're not able to change that to `Connect to Twitter` then please change the description here from `Connect to Twitter` to `Save Changes`.

**Comment by jeffpaul** [comment]
1. When first navigating into the settings screen, are we able to hide / not show the `Close connection settings ^` link?  That would be ideal as a new user will need to complete these steps for Autoshare to work.  After they complete those fields and click `Save Changes`, then the `Open connection settings \/` and `Close connection settings ^` links make sense.

1. Please update the settings page title from `Autoshare for Twitter` to `Autoshare for Twitter Settings`.

**Comment by dinhtungdu** [comment]
@jeffpaul Thanks for your reviews, I addressed your comments, please take another look!

**Comment by jeffpaul** [comment]
Nice work @dinhtungdu, looks good to me.  Would be ideal to get a quick CR from @johnwatkins0 before merging.

**Comment by johnwatkins0** [comment]
I'll take a look and do some quick QA as well.

**Review by johnwatkins0 — CHANGES_REQUESTED** [review]
This is looking great and works well in my testing. Left a few comments, but all little things.

File: assets/js/admin-autoshare-for-twitter-settings.js [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+	} );
+
+	/**
+	 * Get closest Element.
+	 *
+	 * @param {HTMLElement} el Element to get parent.
+	 * @param {string} selector CSS selector to match.
+	 */
+	function getClosest( el, selector ) {
```
[the remark this answers is not in the document]
Comment (dinhtungdu) [reply assets/js/admin-autoshare-for-twitter-settings.js]: Ops, forgot to delete it after changing the logic

File: includes/admin/settings.php [hunk 9]
```diff
@@ -113,12 +202,157 @@ function register_settings() {
...
+ * Helper for ouputing credentials section.
+ *
+ * @return void
+ */
+function cred_section_cb() {
+	?>
+	<section class="credentials-setup <?php echo Utils\is_twitter_configured() ? 'connected' : ''; ?>">
+		<?php if ( Utils\is_twitter_configured() ) : ?>
+		<p class="credentials-actions">
+			<a href="JavaScript:void(0);" class="open">
```
[the remark this answers is not in the document]
Comment (dinhtungdu) [reply includes/admin/settings.php]: because I want the markup match with the visual, see: https://prnt.sc/rbsj5t

**Comment by dinhtungdu** [comment]
@johnwatkins0 Thanks for your detailed review! This PR is ready to review again.

**Review by johnwatkins0 — APPROVED** [review]
Looks good!

**Pull request merged**